src_plugins/show/ppx_deriving_show.cppo.ml

```diff
@@ -317,14 +312,36 @@ let str_of_type ~options ~path ({ ptype_loc = loc } as type_decl) =
          (Ppx_deriving.sanitize ~quoter (polymorphize prettyprinter));
    Vb.mk ~attrs:[no_warn_32] (Pat.constraint_ show_var show_type) (polymorphize stringprinter);]
 
-let () =
-  let loc = !Ast_helper.default_loc in
-  Ppx_deriving.(register (create deriver
-    ~core_type: (Ppx_deriving.with_quoter (fun quoter typ ->
-      [%expr fun x -> Ppx_deriving_runtime.Format.asprintf "%a" (fun fmt -> [%e expr_of_typ quoter typ]) x]))
-    ~type_decl_str: (fun ~options ~path type_decls ->
-      [Str.value Recursive (List.concat (List.map (str_of_type ~options ~path) type_decls))])
-    ~type_decl_sig: (fun ~options ~path type_decls ->
-      List.concat (List.map (sig_of_type ~options ~path) type_decls))
-    ()
-  ))
+(* TODO: add to ppxlib? *)
+let ebool: _ Ast_pattern.t -> _ Ast_pattern.t =
+  Ast_pattern.map1 ~f:(function
+    | [%expr true] -> true
+    | [%expr false] -> false
+    | _ -> failwith "not bool")
+let args () = Deriving.Args.(empty +> arg "with_path" (ebool __))
+(* TODO: add arg_default to ppxlib? *)
+
+let impl_generator = Deriving.Generator.V2.make (args ()) (fun ~ctxt (_, type_decls) with_path ->
+  let path =
+    (* based on https://github.com/thierry-martinez/ppx_show/blob/db00365470bcbf602d931c1bfd155be459379c5c/src/ppx_show.ml#L384-L387 *)
+    let code_path = Ppxlib.Expansion_context.Deriver.code_path ctxt in
+    (* main_module_name contains ".cppo" due to #line directives? *)
+    (* Ppxlib.Code_path.(main_module_name code_path :: submodule_path code_path) *)
+    Ppxlib.Code_path.(Ppx_deriving.module_from_input_name () @ submodule_path code_path)
```

Comment:
Getting a compatible path turns out to be hacky, is there anything better to do?

Reply:
I'm no cppo user myself and prefer other tools for the job but from what I understand, `.cppo` files are not valid ocaml files and therefore the ppxlib driver can't deal with a .cppo.ml(i) file. If it ends up with one it's probably due to an error in the compilation chain isn't it?

I guess it could happen that the filename contains the extension if someone was using cppo via `-pp` and the ppx driver via `-ppx` but I don't think this is the most common case, especially not for dune users.

Could you describe in what context this happened?

Reply:
The problem is that code transformations (like cppo) can turn `foo.cppo.ml` into `foo.ml` but it adds directives like `#line "foo.cppo.ml"` (or whatever the exact syntax). So even when tools run on the produced `foo.ml`, they use overridden locations from these directives to instead output errors, etc in the original user code.

The issue in ppxlib (https://github.com/ocaml-ppx/ppxlib/issues/403) is that its `main_module_name` will be `Foo.cppo` in such case (with the `.ml` suffix removed), which isn't a valid OCaml module name. Instead we just want `Foo` (because it's `foo.ml` after all), so there's a workaround.

Reply:
Right, I see. I'll look into this on the ppxlib side. This doesn't change anything for this PR atm though.

We can adapt the code once the fix is released in ppxlib!